README.md

```diff
@@ -1,2 +1,193 @@
+```
+
+#### Encrypt credentials at rest
+
+Credentials (token signatures, passwords and secrets) are always encrypted at rest.
+
+#### Implement peer reviewed IETF Standards
+
+Fosite implements [rfc6749](https://tools.ietf.org/html/rfc6749) and enforces countermeasures suggested in [rfc6819](https://tools.ietf.org/html/rfc6819).
+
+### Provide extensibility and interoperability
+
+... because OAuth2 is an extensible and flexible **framework**. Fosite let's you register new response types, new grant
+types and new response key value pares. This is useful, if you want to provide OpenID Connect on top of your
+OAuth2 stack. Or custom assertions, what ever you like and as long as it is secure. ;)
+
+## Usage
+
+This section is WIP and we welcome discussions via PRs or in the issues.
+
+### Store
+
+To use fosite, you need to implement `fosite.Storage`. Example implementations (e.g. postgres) of `fosite.Storage`
+will be added in the close future.
+
+### Authorize Endpoint
+
+```go
+package main
+
+import(
+    "github.com/ory-am/fosite"
+    "github.com/ory-am/fosite/session"
+    "github.com/ory-am/fosite/storage"
+	"golang.org/x/net/context"
+)
+
+// Let's assume that we're in a http handler
+func handleAuth(rw http.ResponseWriter, req *http.Request) {
+    store := fosite.NewPostgreSQLStore()
+    oauth2 := fosite.NewDefaultOAuth2(store)
+    ctx := context.Background()
+
+    authorizeRequest, err := oauth2.NewAuthorizeRequest(ctx, r)
+    if err != nil {
+        // ** This part of the API is not finalized yet **
+        // oauth2.RedirectError(rw, error)
+        // oauth2.WriteError(rw, error)
+        // oauth2.handleError...
+        // ****
+        return
+    }
+
+    // you have now access to authorizeRequest, Code ResponseTypes, Scopes ...
+    // and can show the user agent a login or consent page.
+
+    // it would also be possible to redirect the user to an identity provider (google, microsoft live, ...) here
+    // and do fancy stuff like OpenID Connect amongst others
+
+    // Once you have confirmed the users identity and consent that he indeed wants to give app XYZ authorization,
+    // you will use the user's id to create an authorize session
+    user := "12345"
+
+    // NewAuthorizeSessionSQL uses gob.encode to safely store data set with SetExtra
+    session := fosite.NewAuthorizeSessionSQL(authorizeRequest, user)
+
+    // You can store additional metadata, for example:
+    session.SetExtra(map[string]interface{}{
+         "userEmail": "foo@bar",
+         "lastSeen": new Date(),
+         "usingIdentityProvider": "google",
+    })
+
+    // or
+    session.SetExtra(&struct{
+        Name string
+    } {
+        Name: "foo"
+    })
+
+    // Now is the time to handle the response types
+
+    // you can use a custom list of response type handlers by setting
+    // oauth2.ResponseTypeHandlers = []fosite.ResponseTypeHandler{}
+    response, err := oauth2.HandleResponseTypes(ctx, authorizeRequest, r)
```

Comment:
The question here is, if we should pass the session as well to HandleResponseTypes. This would enable ResponseTypeHandlers to modify the session and persist data for the grant request (usually /token endpoint). Of course, ResponseTypeHandlers are also able to use their own stores (e.g. OpenIDHandler uses OpenIDConnectStore), which might be more reasonable and would reduce naming conflicts etc.